bugtool/cmd/root.go

```diff
@@ -0,0 +1,237 @@
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/vishvananda/netlink"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+// RootCmd is the top level command for the bugtool.
+var RootCmd = &cobra.Command{
+	Use:   "bugtool",
+	Short: "Cilium agent debugging tool",
+	Long:  "cilium-bugtool - retrieve information and files for debugging single Cilium node",
+	Run: func(cmd *cobra.Command, args []string) {
+		common.RequireRootPrivilege("bugtool")
+		runTool()
+	},
+}
+
+var (
+	serve      bool
+	port       int
+	dumpPath   string
+	host       string
+	disclaimer = `DISCLAIMER
+This tool has copied information about your environment.
+If you are going to register a issue on GitHub, please
+only provide files from the archive you have reviewed
+for sensitive information. For more information on what
+is being collected, please see[0].
+
+Thanks.
+
+[0]: http://docs.cilium.io/en/stable/troubleshooting/#debugtool
+`
+)
+
+func init() {
+	RootCmd.Flags().BoolVar(&serve, "serve", false, "Start HTTP server to serve static files")
+	RootCmd.Flags().IntVarP(&port, "port", "p",
+		4444, "Port to use for the HTTP server, (default 4444)")
+	RootCmd.Flags().StringVarP(&dumpPath, "tmp", "t", "/tmp/", "Path to store extracted files")
+	RootCmd.Flags().StringVarP(&host, "host", "H", "", "URI to server-side API")
+}
+
+func runTool() {
+	defer fmt.Println(disclaimer)
+
+	// Prevent collision with other directories
+	dbgDir, err := ioutil.TempDir(dumpPath, "cilium-bugtool-")
+	if err != nil {
+		log.WithError(err).Fatal("failed to create debug directory")
+	}
+
+	copyStateDir(dbgDir)
+	copySystemCmds(createDir(dbgDir, "cmd"))
+	copyCiliumCmds(createDir(dbgDir, "cilium"))
+
+	confDir := createDir(dbgDir, "conf")
+	resolveFilePath := filepath.Join(confDir, "resolv.conf")
+	os.Create(resolveFilePath)
+	copyFile("/etc/resolv.conf", resolveFilePath)
+	copyKernelConfig(confDir)
+
+	createArchive(dbgDir)
+
+	if serve {
+		serveStaticFiles(dbgDir)
+	}
+
+	if err := os.RemoveAll(dbgDir); err != nil {
```

Comment:
I'm not sure, but I think ListenAndServe never returns (in serveStaticFiles) and if that's the case then the only way to exit would be with a ctrl-C or something which then would mean we never run this cleanup step. Not that big a deal since these are in temp directories.
I think there are ListenAndServe wrappers or versions that allow exit. If not, we can run that in a goroutine, and wait here for some other trigger (possibly a os.Signal)

Reply:
Good catch. Did not notice it would skip it. For the temporary files as you point it's not a big issue but we should always print the disclaimer. Changing it to defer by default but also support `SIGINT`.